File: Javascript/src/char-processors.js
Set.prototype.isSubsetOf = function (otherSet) {
    return [...this].map(e => otherSet.has(e))
                    .every(cond => cond === true);
}

const digitsByLanguage = {
    "bangla": ['০', '১', '২', '৩', '৪', '৫', '৬', '৭', '৮', '৯'],
    "english": ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
}

export function findOutLanguage(numStr) {
    const numStrDigitSet = new Set([...numStr]);
    numStrDigitSet.delete('.');
    numStrDigitSet.delete('+');
    numStrDigitSet.delete('-');

    for (const language in digitsByLanguage) {
        const digitSet = new Set(digitsByLanguage[language]);
        if (numStrDigitSet.isSubsetOf(digitSet))
            return language;
    }

    throw new Error(`supported languages ${Object.keys(digitsByLanguage)}`);
}

export function convertString(numStr, fromLanguage, toLanguage) {
    numStr = String(numStr);
    return [...numStr].map(char => convertCharacter(char, fromLanguage, toLanguage))
                      .join('');
}

function convertCharacter(char, fromLanguage, toLanguage) {
    const digitIndex = digitsByLanguage[fromLanguage].indexOf(char);
    if (digitIndex < 0) return char;
    return digitsByLanguage[toLanguage][digitIndex];
}

export function convertIndex(index, language) {
    const digitAsInt = Number(index);

    if (!isDigit(digitAsInt)) throw new Error('Must be a digit [0,9]');

    return digitsByLanguage[language][index];
}

function isDigit(digitAsInt) {
    return 0 <= digitAsInt && digitAsInt <= 9;
}

File: Python/src/char-processors.py
from constants import BANGLA, ENGLISH

digits_by_language = {
    BANGLA: ['০', '১', '২', '৩', '৪', '৫', '৬', '৭', '৮', '৯'],
    ENGLISH: ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
}


def find_out_language(num_str):
    num_str_digit_set = set(num_str)
    num_str_digit_set.discard('.')
    num_str_digit_set.discard('+')
    num_str_digit_set.discard('-')

    for language, digits in digits_by_language.items():
        digit_set = set(digits)
        if num_str_digit_set.issubset(digit_set):
            return language

    raise Exception(f'supported languages {digits_by_language.keys()}')


def convert_string(num_str: str, from_language: str, to_language: str) -> str:
    return ''.join(map(lambda char: convert_character(char, from_language, to_language), num_str))


def convert_character(char: str, from_language: str, to_language: str) -> str:
    try:
        digit_index = digits_by_language[from_language].index(char)
    except ValueError:
        return char

    return digits_by_language[to_language][digit_index]


def convert_index(index: int, language: str) -> str:
    if not is_digit(index):
        raise Exception('Must be a digit [0,9]')

    return digits_by_language[language][index]


def is_digit(digit: int) -> bool:
    return 0 <= digit <= 9
